File: 32--Save-the-Clock-Tower/HighVoltage/High Voltage/ElectricityConversions.m
//
//  CalculatorBrain+Extras.m
//  High Voltage
//
//  Created by Ben Gohlke on 3/10/15.
//  Copyright (c) 2015 The Iron Yard. All rights reserved.
//

#import "ElectricityConversions.h"

@implementation ElectricityConversions

- (instancetype)init
{
    self = [super init];
    if (self)
    {
        _allValuesFound = NO;
    }
    
    return self;
}

- (void)findOtherValuesIfPossible
{
    int count = 0;
    if (self.ampsString && ![self.ampsString isEqualToString:@""])
    {
        count++;
    }
    if (self.ohmsString && ![self.ohmsString isEqualToString:@""])
    {
        count++;
    }
    if (self.voltsString && ![self.voltsString isEqualToString:@""])
    {
        count++;
    }
    if (self.wattsString && ![self.wattsString isEqualToString:@""])
    {
        count++;
    }
    
    if (count >= 2)
    {
        [self calculateAllValues];
    }
}

- (void)calculateAllValues
{
    float amps = 0, ohms = 0, volts = 0, watts = 0;
    
    if (self.ohmsString && self.voltsString)
    {
        ohms = [self.ohmsString floatValue];
        volts = [self.voltsString floatValue];
        amps = volts / ohms;
        watts = volts * amps;
    }
    else if (self.ohmsString && self.ampsString)
    {
        amps = [self.ampsString floatValue];
        ohms = [self.ohmsString floatValue];
        volts = amps * ohms;
        watts = volts * amps;
    }
    else if (self.ohmsString && self.wattsString)
    {
        ohms = [self.ohmsString floatValue];
        watts = [self.wattsString floatValue];
        amps =  sqrtf(watts / ohms);
        volts = amps * ohms;
    }
    else if (self.ampsString && self.voltsString)
    {
        amps = [self.ampsString floatValue];
        volts = [self.voltsString floatValue];
        ohms = volts / amps;
        watts = volts * amps;
    }
    else if (self.ampsString && self.wattsString)
    {
        amps = [self.ampsString floatValue];
        watts = [self.wattsString floatValue];
        volts = watts / amps;
        ohms = volts / amps;
    }
    else if (self.voltsString && self.wattsString)
    {
        volts = [self.voltsString floatValue];
        watts = [self.wattsString floatValue];
        amps = watts / volts;
        ohms = volts / amps;
    }
    
    self.voltsString = [NSString stringWithFormat:@"%g", volts];
    self.ampsString = [NSString stringWithFormat:@"%g", amps];
    self.ohmsString = [NSString stringWithFormat:@"%g", ohms];
    self.wattsString = [NSString stringWithFormat:@"%g", watts];
    
    self.allValuesFound = YES;
    
    [self.delegate valuesWereCalculated];
}

@end

File: 16--Come-About/SwiftVoltage/High Voltage/ElectricityConversions.swift
//
//  ElectricityConversions.swift
//  High Voltage
//
//  Created by Ben Gohlke on 7/26/15.
//  Copyright (c) 2015 The Iron Yard. All rights reserved.
//

import Foundation

class ElectricityConversions
{
    var allValuesFound: Bool
    var delegate: ElectricityConversionsDelegate?
    
    var ampsString: String
    var ohmsString: String
    var wattsString: String
    var voltsString: String
    
    init()
    {
        allValuesFound = false
        ampsString = ""
        ohmsString = ""
        wattsString = ""
        voltsString = ""
    }
    
    func findOtherValuesIfPossible()
    {
        var count = 0
        if ampsString != ""
        {
            count++
        }
        if ohmsString != ""
        {
            count++
        }
        if wattsString != ""
        {
            count++
        }
        if voltsString != ""
        {
            count++
        }
        if count >= 2
        {
            calculateAllValues()
        }
        
    }
    
    private func calculateAllValues()
    {
        var amps = 0.0, ohms = 0.0, volts = 0.0, watts = 0.0
        if ohmsString != "" && voltsString != ""
        {
            ohms = (ohmsString as NSString).doubleValue
            volts = (voltsString as NSString).doubleValue
            amps = volts / ohms
            watts = volts * amps
        }
        else if ohmsString != "" && ampsString != ""
        {
            ohms = (ohmsString as NSString).doubleValue
            amps = (ampsString as NSString).doubleValue
            volts = amps * ohms
            watts = volts * amps
        }
        else if ohmsString != "" && wattsString != ""
        {
            ohms = (ohmsString as NSString).doubleValue
            watts = (wattsString as NSString).doubleValue
            amps = sqrt(watts / ohms)
            volts = amps * ohms
        }
        else if ampsString != "" && voltsString != ""
        {
            volts = (voltsString as NSString).doubleValue
            amps = (ampsString as NSString).doubleValue
            ohms = volts / amps
            watts = volts * amps
        }
        else if wattsString != "" && ampsString != ""
        {
            watts = (wattsString as NSString).doubleValue
            amps = (ampsString as NSString).doubleValue
            volts = watts / amps
            ohms = volts / amps
        }
        else if voltsString != "" && wattsString != ""
        {
            volts = (voltsString as NSString).doubleValue
            watts = (wattsString as NSString).doubleValue
            amps = watts / volts
            ohms = volts / amps
        }
        
        voltsString = "\(volts)"
        ampsString = "\(amps)"
        ohmsString = "\(ohms)"
        wattsString = "\(watts)"
        
        allValuesFound = true
        delegate?.valuesWereCalculated()
    }
}